### PR TITLE
Feature/redux

### PR DESCRIPTION
**Pull request by atharateeq-mergestack**
# Description

Implement redux with saga in current system.
Made code refactoring on dashboard component.
Code refactoring is made by keep container implementation in mind. 

## Type of change

- [ ] Bug fix 
- [ ] New feature 
- [X] Resolve PR changes 
- [ ] Breaking change
- [ ] This change requires a documentation update

### CODE REVIEW

File: src/components/DateTimeDisplay.tsx [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { formatDateTime } from 'utils/dateFormatter';
+
+interface DateTimeDisplayProps {
+  date: string;
+};
+
+const DateTimeDisplay = ({ date } : DateTimeDisplayProps) => {
+  const { formattedDate, formattedTime } = formatDateTime(date);
```
Comment (JavedJutt) [inline src/components/DateTimeDisplay.tsx]: is there a specific format in which we are converting date and name function as that

File: src/components/Modal/index.tsx [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import { useDispatch } from 'react-redux';
+
+import { IResultData } from 'utils/types';
+import { deleteResultRequest } from 'reduxStore/actions/resultActions';
+import 'components/Modal/style.css'; 
+
+interface ModalProps{
```
Comment (JavedJutt) [inline src/components/Modal/index.tsx]: pls put I before any interface `ModalProps` to `IModalProps `

File: src/components/NoRecord/index.tsx [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import 'components/NoRecord/style.css'
+function NoRecord (){
```
Comment (JavedJutt) [inline src/components/NoRecord/index.tsx]: always use es6 format

File: src/components/Table/TableRow.tsx [hunk 4]
```diff
@@ -1,49 +1,72 @@
...
+  const handleActionClick = (student: IResultData) => {
+    setSelectedStudent(student);
+    setDropdownVisible(!dropdownVisible);
+    if (dropdownRef.current) {
+      const dropdownRect = dropdownRef.current.getBoundingClientRect();
+      if (dropdownRect.bottom > window.innerHeight) {
+        dropdownRef.current.style.top = `-${dropdownRect.height}px`;
+      } else {
+        dropdownRef.current.style.top = `100%`;
+      }
```
Comment (JavedJutt) [inline src/components/Table/TableRow.tsx]: what's purpose of this? is drop down height creating issue in responsiveness?

File: src/components/Table/index.tsx [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import TableHeader from 'components/Table/TableHeader';
+import NoRecord from 'components/NoRecord';
+import { RootState } from 'store';
+import { fetchResultsRequest } from 'reduxStore/actions/resultActions';
+
+
+const Table = () => {
+  const dispatch = useDispatch();
+ 
+  const { results, fetched } = useSelector((state : RootState) => state.results);
```
Comment (JavedJutt) [inline src/components/Table/index.tsx]: you haven't implemented containers yet?

File: src/pages/AddResult/index.tsx [hunk 6]
```diff
@@ -1,68 +1,48 @@
...
-      }
-    };
-
-    fetchData();
-  }, [existingData, setValue]);
+  useEffect(() => {  
+    if(!fetched)  {
+      dispatch(fetchSubjectsRequest());
+    }
+  }, [dispatch, fetched, subjects]);
```
Comment (JavedJutt) [inline src/pages/AddResult/index.tsx]: we need to use containers

File: src/components/SummaryCards/SummaryCard.tsx [hunk 7]
```diff
@@ -1,12 +1,6 @@
...
+import { ISummaryCardProps } from "utils/types";
 
-interface SummaryCardProps {
-  title: string;
-  value: string | number | undefined;
-  className: string;
-}
-
-const SummaryCard = ({ title, value, className } : SummaryCardProps) => {
+const SummaryCard = ({ title, value, className } : ISummaryCardProps) => {
```
Comment (JavedJutt) [inline src/components/SummaryCards/SummaryCard.tsx:3]: why did we removed interface from here?

File: src/containers/index.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+import SummaryCards from 'components/SummaryCards';
+import Table from 'components/Table';
+import AddResult from 'pages/AddResult';
+import Subject from 'pages/Subject';
+
+import { IResultData, ICreateResult, ICreateSubject } from 'utils/types';
+import { deleteResultAction, fetchResultsAction, createResultAction, updateResultAction } from 'store/actions/result';
+import { fetchSubjectsAction, createSubjectAction } from 'store/actions/subject';
+import { selectCardData } from './selectors/result';
+import { selectSubjectDropDown } from './selectors/subject';
```
Comment (JavedJutt) [inline src/containers/index.ts]: imports

File: src/containers/index.ts [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+  loading: state.subjects.loading,
+  error: state.subjects.error
+});
+;
+
+const mapDispatchToPropsSubject = (dispatch: ThunkDispatch<RootState, unknown, AnyAction>) => ({
+  createSubject: (data: ICreateSubject) => {dispatch(createSubjectAction(data))},
+});
+
+const SubjectContainer = connect(mapStateToPropsSubject, mapDispatchToPropsSubject)(Subject);
```
Comment (JavedJutt) [inline src/containers/index.ts]:
```suggestion
export const SubjectContainer = connect(mapStateToPropsSubject, mapDispatchToPropsSubject)(Subject);
```

File: src/store/actions/result/index.ts
Comment (JavedJutt) [inline src/store/actions/result/index.ts:1]: remove this dispatch logic from actions

File: src/utils/dateFormatter.ts [hunk 10]
```diff
@@ -0,0 +1,13 @@
+import { format } from 'date-fns';
```
Comment (JavedJutt) [inline src/utils/dateFormatter.ts]: we can use moment js also

File: src/components/SummaryCards/SummaryCard.tsx (hunk 7 above)
Comment (atharateeq-mergestack) [reply src/components/SummaryCards/SummaryCard.tsx:3]: I have not remove the interface but update name by adding I in start of interface name.

**Review by JavedJutt — APPROVED** [review]
LGTM

**Pull request merged**